www/js/crypter/crypto_app.js

```diff
@@ -51,11 +53,45 @@ window.filesender.crypto_app = function () {
             v2019_generated_password_that_is_full_256bit: 2
         },
 
+        createChunkIDArray: function( chunkid ) {
+            var ret = new Uint8Array(4);
+            ret[0] = chunkid>>0  & 0xFF;
+            ret[1] = chunkid>>8  & 0xFF;
+            ret[2] = chunkid>>16 & 0xFF;
+            ret[3] = chunkid>>24 & 0xFF;
+            return ret;
+        },
+
+        /**
+         * Create and return an IV of 16 octets (128 bits) constructed as follows:
+         *    12 octets of entropy
+         *     4 octets containing the chunkid
+         *
+         * This is based on Page 19 of OpenFortress 2018 document:
+         *   "The suggested procedure for the case of FileSender is to combine 96 bits
+         *    of random material with a 32-bit chunk counter to form a 128-bit IV."
+         */
+        createIVGCM: function( chunkid ) {
+            var $this = this;
+
+            // 96 bits of entropy
+            ivrandom  = crypto.getRandomValues(new Uint8Array(12));
+
+            // 32 bits of counter from chunkid
+            ivcounter = $this.createChunkIDArray(chunkid);
+
+            // merge these into return value
+            iv = new Uint8Array(16);
+            iv.set(ivrandom);
+            iv.set(ivcounter, ivrandom.length );
+            
+            return iv;
+        },
         
         generateVector: function () {
             return crypto.getRandomValues(new Uint8Array(16));
         },
-        generateKey: function (encryption_details, callback) {
+        generateKey: function (chunkid, encryption_details, callback) {
```

Comment:
There is a problem in mapping the (upload-termed) chunk to a (GCM-termed) chunk counter.  This is probably a mistake or unclarity in my 2018 document that I need to correct.

GCM uses the IV as a counter *of 16-byte chunks* and so, if you upload a larger sequence of bytes, you should change the counter to reflect the number of 16-byte chunks instead of incrementing by one.  So, if you upload 4 kB chunks, then your chunk counter in the IV should increment by ceil(4096/16) = 256 on each round.  (I would only accept multiples of 16 bytes by the way, to avoid trouble on that end, though signatures would catch out any problems with that.)

Perhaps we should find a way that I can trace this logic through your code, to help you do this under the constraints of the encryption algorithms used.

Reply:
The chunkid passed in the above cited code is the floor( offset / chunksize ) calculated in the below:
https://github.com/filesender/filesender/blob/master/www/js/terasender/terasender.js#L116

That is eventually passed to createIVGCM() to make an IV.
https://github.com/filesender/filesender/blob/master/www/js/crypter/crypto_app.js#L74

Setting aside me moving the 96 bits of random material to the same 96 bits for a file. The chunkid is encoded by createIVGCM() using createChunkIDArray().
https://github.com/filesender/filesender/blob/master/www/js/crypter/crypto_app.js#L56

At the moment a chunk of 0 will be encoded as 0
A chunk of 1 (4mb into the file) will be encoded as 1
and so on.

To clarify, after moving to chunkid*ceil(chunk-size-in-bytes / 16) from above we should encode 262144 when passed chunkid=1 by FileSender. And for chunkid=2 then 524288 is encoded. And the max_file_size is limited as per your above limit.

Reply:
Yes to your clarification!  Multiply the chunkid (0, 1, 2...) by 262144 (0, 262144, 524288, ...) to form the 32-bit counter value.  Indeed, under the max_file_size being 68719476735 or 16 * 2^32 - 1.